### PR TITLE
Make directories in Playlist component per process model

### PR DESCRIPTION
**Pull request by sangwoo108**
The source files should be organized by process model.

```
If a file should be used only by renderer   => renderer/
                         only by browser    => browser/
          can be used by multiple processes => common/
```

More specifically, 
In `common` : mojom, buildflag, features
In `renderer` : PlaylistRenderFrameObserver
In `browser` : all others.

This PR doesn't intend any behavioral changes.

Resolves https://github.com/brave/brave-browser/issues/27507

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by sangwoo108** [comment]
@deeppandya , please note that the mojo file is moving to //brave/components/playlist/common/ .
I believe that's the only thing that affect your work.

**Review by simonhong — APPROVED** [review]
++

File: test/BUILD.gn [hunk 1]
```diff
@@ -873,8 +873,8 @@ test("brave_browser_tests") {
     "//brave/components/l10n/common",
     "//brave/components/l10n/common:test_support",
     "//brave/components/ntp_background_images/buildflags",
-    "//brave/components/playlist",
-    "//brave/components/playlist/buildflags",
+    "//brave/components/playlist/browser",
+    "//brave/components/playlist/common/buildflags",
```
Comment (goodov) [inline test/BUILD.gn:877]: this is incorrect. you should never have `buildflags` and the target it's protecting in the same list.

File: test/BUILD.gn [hunk 2]
```diff
@@ -873,8 +873,8 @@ test("brave_browser_tests") {
     "//brave/components/l10n/common",
     "//brave/components/l10n/common:test_support",
     "//brave/components/ntp_background_images/buildflags",
-    "//brave/components/playlist",
-    "//brave/components/playlist/buildflags",
+    "//brave/components/playlist/browser",
```
Comment (goodov) [inline test/BUILD.gn:876]: `if (enable_playlist)`

File: renderer/sources.gni [hunk 3]
```diff
@@ -46,7 +46,7 @@ if (enable_speedreader) {
 
 if (enable_playlist) {
   brave_chrome_renderer_deps += [
-    "//brave/components/playlist:playlist_features",
+    "//brave/components/playlist/common:features",
     "//brave/components/playlist/renderer",
```
Comment (goodov) [inline renderer/sources.gni:50]: `//brave/components/playlist/renderer` and `//brave/components/playlist/browser` should depend publicly on `//brave/components/playlist/common`, then you only need to depend on `browser` or `renderer` targets (without explicit `common`).

File: browser/playlist/test/BUILD.gn [hunk 4]
```diff
@@ -3,7 +3,7 @@
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
 # You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-import("//brave/components/playlist/buildflags/buildflags.gni")
+import("//brave/components/playlist/common/buildflags/buildflags.gni")
 
 source_set("browser_tests") {
   assert(enable_playlist)
```
Comment (goodov) [inline browser/playlist/test/BUILD.gn:9]: pls move this to the top level like so:
```
import("//brave/components/playlist/common/buildflags/buildflags.gni")

assert(enable_playlist)

source_set("browser_tests") {
```

and remove it from `source_set("unit_tests")`

File: components/playlist/common/buildflags/buildflags.gni [hunk 5]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) 2021 The Brave Authors. All rights reserved.
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
-# You can obtain one at http://mozilla.org/MPL/2.0/.
+# You can obtain one at https://mozilla.org/MPL/2.0/.
 
 import("//build/config/ui.gni")
```
Comment (goodov) [inline components/playlist/common/buildflags/buildflags.gni:7]: I assume `enable_playlist_webui` should not be enabled if `enable_playlist` is disabled. If it's the case, then this should be declared like so:
```
declare_args() {
  enable_playlist = toolkit_views || is_android
}

declare_args() {
  enable_playlist_webui = enable_playlist && toolkit_views
}

assert(!enable_playlist_webui || enable_playlist, "Playlist buildflag should be enabled first")
```

File: components/sidebar/BUILD.gn [hunk 6]
```diff
@@ -39,7 +39,7 @@ source_set("unit_tests") {
   deps = [
     "//base",
     "//base/test:test_support",
-    "//brave/components/playlist:playlist_features",
+    "//brave/components/playlist/common:features",
```
Comment (goodov) [inline components/sidebar/BUILD.gn:42]: `//brave/components/playlist/common`

File: components/playlist/common/BUILD.gn [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+
+source_set("features") {
+  assert(enable_playlist)
+
+  sources = [
+    "features.cc",
+    "features.h",
+  ]
+
+  deps = [ "//base" ]
```
Comment (goodov) [inline components/playlist/common/BUILD.gn:16]:
```
public_deps = [
  "//brave/components/playlist/common/buildflags",
  "//brave/components/playlist/common/mojom",
]
```

File: components/playlist/browser/BUILD.gn [hunk 8]
```diff
@@ -1,23 +1,12 @@
...
-    "features.cc",
-    "features.h",
-  ]
-
-  deps = [ "//base" ]
-}
-
-source_set("playlist") {
+source_set("browser") {
   assert(enable_playlist)
```
Comment (goodov) [inline components/playlist/browser/BUILD.gn:10]: same, pls put this on top level

File: components/playlist/browser/resources/BUILD.gn [hunk 9]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) 2021 The Brave Authors. All rights reserved.
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
-# You can obtain one at http://mozilla.org/MPL/2.0/.
+# You can obtain one at https://mozilla.org/MPL/2.0/.
 
 import("//brave/components/common/typescript.gni")
```
Comment (goodov) [inline components/playlist/browser/resources/BUILD.gn:7]: `assert(enable_playlist)`

File: components/playlist/renderer/BUILD.gn [hunk 10]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) 2022 The Brave Authors. All rights reserved.
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
-# you can obtain one at http://mozilla.org/MPL/2.0/.
+# You can obtain one at https://mozilla.org/MPL/2.0/.
```
Comment (goodov) [inline components/playlist/renderer/BUILD.gn:5]: `assert(enable_playlist)`

File: test/BUILD.gn [hunk 11]
```diff
@@ -1025,7 +1023,10 @@ test("brave_browser_tests") {
   }
 
   if (enable_playlist) {
-    deps += [ "//brave/browser/playlist/test:browser_tests" ]
+    deps += [
+      "//brave/browser/playlist/test:browser_tests",
+      "//brave/components/playlist/browser",
```
Comment (goodov) [inline test/BUILD.gn:1028]: this looks weird. it should go into public_deps of `//brave/browser/playlist/test:browser_tests`.

File: components/playlist/renderer/BUILD.gn [hunk 12]
```diff
@@ -10,11 +14,12 @@ source_set("renderer") {
...
   deps = [
-    "//brave/components/playlist:playlist_features",
     "//content/public/renderer",
     "//third_party/blink/public:blink_headers",
     "//third_party/blink/public/common",
     "//url",
     "//v8",
   ]
+
+  public_deps = [ "//brave/components/playlist/common" ]
```
Comment (goodov) [inline components/playlist/renderer/BUILD.gn:24]: https://gn.googlesource.com/gn/+/main/docs/style_guide.md#ordering-within-a-target

File: components/playlist/common/mojom/BUILD.gn [hunk 13]
```diff
@@ -1,7 +1,7 @@
 # Copyright (c) 2022 The Brave Authors. All rights reserved.
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this file,
-# you can obtain one at http://mozilla.org/MPL/2.0/.
+# You can obtain one at https://mozilla.org/MPL/2.0/.
 
 import("//mojo/public/tools/bindings/mojom.gni")
```
Comment (goodov) [inline components/playlist/common/mojom/BUILD.gn:7]: maybe also assert here?

File: components/sidebar/BUILD.gn [hunk 14]
```diff
@@ -22,7 +22,7 @@ static_library("sidebar") {
     "//base",
     "//brave/components/brave_wallet/common:common_utils",
     "//brave/components/l10n/common",
-    "//brave/components/playlist:playlist_features",
+    "//brave/components/playlist/common",
```
Comment (goodov) [inline components/sidebar/BUILD.gn:25]: should be under buildflag

File: components/sidebar/BUILD.gn [hunk 15]
```diff
@@ -39,7 +39,7 @@ source_set("unit_tests") {
   deps = [
     "//base",
     "//base/test:test_support",
-    "//brave/components/playlist:playlist_features",
+    "//brave/components/playlist/common",
```
Comment (goodov) [inline components/sidebar/BUILD.gn:42]: should be under buildflag

File: components/sidebar/sidebar_service.cc [hunk 16]
```diff
@@ -21,7 +21,7 @@
 #include "base/values.h"
 #include "brave/components/brave_wallet/common/common_util.h"
 #include "brave/components/l10n/common/localization_util.h"
-#include "brave/components/playlist/features.h"
+#include "brave/components/playlist/common/features.h"
```
Comment (goodov) [inline components/sidebar/sidebar_service.cc:24]: should be under buildflag

File: components/sidebar/sidebar_service_unittest.cc [hunk 17]
```diff
@@ -9,7 +9,7 @@
 #include "base/containers/contains.h"
 #include "base/ranges/algorithm.h"
 #include "base/test/scoped_feature_list.h"
-#include "brave/components/playlist/features.h"
+#include "brave/components/playlist/common/features.h"
```
Comment (goodov) [inline components/sidebar/sidebar_service_unittest.cc:12]: under buildflag

File: browser/ui/BUILD.gn [hunk 18]
```diff
@@ -764,10 +764,8 @@ source_set("ui") {
       "webui/playlist_ui.cc",
       "webui/playlist_ui.h",
     ]
-    deps += [
-      "//brave/components/playlist/mojom",
-      "//brave/components/playlist/resources:generated_resources",
-    ]
+    deps +=
+        [ "//brave/components/playlist/browser/resources:generated_resources" ]
```
Comment (goodov) [inline browser/ui/BUILD.gn:768]: `playlist_ui.h/cc` includes playlist mojom and `playlist_service.h`, it means you also need `//brave/components/playlist/browser` here.

this `"ui"` target has `check_includes = false`, that's why gn check doesn't show anything here :/

File: components/sidebar/sidebar_service.cc [hunk 19]
```diff
@@ -34,6 +33,10 @@
 #include "ui/base/l10n/l10n_util.h"
 #include "ui/base/resource/resource_bundle.h"
 
+#if BUILDFLAG(ENABLE_PLAYLIST)
```
Comment (goodov) [inline components/sidebar/sidebar_service.cc:36]: I don't see `playlist/common/buildflags/buildflags.h` include here.

File: components/sidebar/sidebar_service_unittest.cc [hunk 20]
```diff
@@ -20,6 +19,10 @@
 #include "testing/gmock/include/gmock/gmock.h"
 #include "testing/gtest/include/gtest/gtest.h"
 
+#if BUILDFLAG(ENABLE_PLAYLIST)
```
Comment (goodov) [inline components/sidebar/sidebar_service_unittest.cc:22]: `playlist/common/buildflags/buildflags.h`

File: browser/playlist/test/BUILD.gn [hunk 21]
```diff
@@ -17,10 +17,12 @@ source_set("browser_tests") {
...
   ]
 
+  public_deps = [ "//brave/components/playlist/browser" ]
+
   deps = [
     "//base",
-    "//brave/components/playlist",
-    "//brave/components/playlist/mojom",
+    "//brave/components/playlist/browser",
+    "//brave/components/playlist/common/mojom",
```
Comment (goodov) [inline browser/playlist/test/BUILD.gn:25]: these two lines can be removed (they're available via public_deps now).

File: components/sidebar/sidebar_service.cc (hunk 19 above)
Comment (sangwoo108) [reply components/sidebar/sidebar_service.cc:37]: Yeah, that was exactly what I ran into! thanks :)

File: browser/ui/BUILD.gn (hunk 18 above)
Comment (sangwoo108) [reply browser/ui/BUILD.gn]: Than makes sense! Done.

**Review by goodov — APPROVED** [review]

File: components/playlist/browser/BUILD.gn [hunk 22]
```diff
@@ -1,25 +1,14 @@
...
-  ]
-
-  deps = [ "//base" ]
-}
-
-source_set("playlist") {
-  assert(enable_playlist)
+assert(enable_playlist)
 
+source_set("browser") {
```
Comment (bridiver) [inline components/playlist/browser/BUILD.gn:11]: please change to `static_library` and we should add a follow-up issue to change to `component`

**Review by bridiver — APPROVED** [review]

File: components/playlist/browser/BUILD.gn (hunk 22 above)
Comment (sangwoo108) [reply components/playlist/browser/BUILD.gn:11]: Good call. I was wondering why we don't use component target whe I joined the team. filed an issue here: https://github.com/brave/brave-browser/issues/27764

**Review by iefremov — APPROVED** [review]

**Pull request merged**